Cosine Similarity and Text Embeddings (#817)
earthly/website: blog/_posts/2024-04-04-cosine_similarity_text_embeddings.md

One way to solve this problem is to create a table of words and their membership in various classes. A dog and a cat are both pets. A brick is a building material, and a shoe is footwear. So if we have a list of bits, marking `isPet`,`isConstruction`,`isMaterial`, then we can store words like this:

~~~{.python caption=""}
items = {
  "dog":[1,0,0],
   "cat":[1,0,0],
   "bulldog":[1,0,0],
   "brick":[0,1,0],
   "shoe":[0,0,1],
}
~~~

Pedants might say, well, someone could have a pet brick, couldn't they? And maybe you could use a shoe as construction material? And yes, that is true. Categories are not all or nothing. Let's make them float from 1 to 0. The closer to 1, the more relevant the word is to the category.

~~~{.python caption=""}
items = {
  "dog":[1.0,0.0,0.0],
   "cat":[1.0,0.0,0.0],
   "bulldog":[1.0,0.0,0.0],
   "brick":[0.1,1.0,0.1],
   "shoe":[0.1,0.1,1.0],
}
~~~

Now, if you look at these three numbers as a point in three-dimensional space, you can see that we've found a way to map a word into three-dimensional space such that items near each other are related to the dimensions we care about.

> In our footwear, construction materials and pets website, we should find that this view of our has three pretty clear clusters of related data, but there might be some out outlier groups for the pet rock people of the world. This projection of the data is a text embedding.

Reviewer: ```suggestion
In our footwear, construction materials, and pets website, we should find that this view of ours has three pretty clear clusters of related data, but there might be some outlier groups for the pet rock people of the world. This projection of the data is a text embedding.
```

> The problem with all this is coming up with all the dimensions and the giant membership list for every word that's important to you. In the real world, we will have a lot more words than this, and we will need a lot more categories to disambiguate them. For instance, many words would score [0,0,0] like `sadness` or `purple` and `philosophy` even though they have nothing to do with each other. We will get to that soon enough, but assuming we have these values, how do we figure out what's related to what?

Reviewer: ```suggestion
The problem with all this is coming up with all the dimensions and the giant membership list for every word that's important to you. In the real world, we will have a lot more words than this, and we will need a lot more categories to disambiguate them. For instance, many words would score `[0,0,0]` like `sadness` or `purple` and `philosophy` even though they have nothing to do with each other. We will get to that soon enough, but assuming we have these values, how do we figure out what's related to what?
```

## What Is Cosine Similarity

> Ok, it's time to get a little mathy. If we take our points in three-dimensional space and treat them as a vector from [0,0,0] to their value, we get a bunch of arrows in three-dimensional space. Here is `dog` and `brick`.

Reviewer: ```suggestion
Ok, it's time to get a little mathy. If we take our points in three-dimensional space and treat them as a vector from `[0,0,0]` to their value, we get a bunch of arrows in three-dimensional space. Here is `dog` and `brick`.
```

    magnitude_b = np.linalg.norm(vector_b)
    cosine_of_angle = dot_product / (magnitude_a * magnitude_b)
    angle_in_degrees = np.degrees(np.arccos(cosine_of_angle))
    return angle_in_degrees

# Example vectors for 'brick' and 'shoe'
vector_brick = np.array([0.1, 1.0, 0.1])
vector_shoe = np.array([0.1, 0.1, 1.0])

calculate_angle_degrees(vector_brick, vector_shoe)
~~~

 ( How we calculate that angle isn't essential if you want to gloss over it. )

 This gives `78.118` degrees for the angle between `brick` and `shoe`. The maximum possible angle with this formula is 90 degrees, completely perpendicular to each other. And the min value is 0 degrees, the two angles are exactly the same.

 To get a similarity score, we just need to invert these values to get them between 0 and 1. 0 degrees should be our exact match value 1, and 90 degrees should be 0. That projection is the cosine of the angle.

~~~{.python caption=""}
 import numpy as np

def cosine_similarity(vector_a, vector_b):
    dot_product = np.dot(vector_a, vector_b)
    magnitude_a = np.linalg.norm(vector_a)
    magnitude_b = np.linalg.norm(vector_b)
    return dot_p

cosine_similarity(vector_brick, vector_shoe)

~~~

> The similarity between `brick` and `shoe` is 0.20. Not very high, corresponding to an angle of ~70 degrees of difference. That is the cosine similarity.

Reviewer: ```suggestion
The similarity between `brick` and `shoe` is `0.20`. Not very high, corresponding to an angle of ~70 degrees of difference. That is the cosine similarity.
```

  8.0413e-01  7.4880e+00  6.1078e+00 -4.2130e+00 -1.5384e-01 -5.4995e+00
  1.0896e+01  3.9278e+00 -1.3601e-01  7.7732e-02  3.2218e+00 -5.8777e+00
  6.1359e-01 -2.4287e+00  6.2820e+00  1.3461e+01  4.3236e+00  2.4266e+00
 -2.6512e+00  1.1577e+00  5.0848e+00 -1.7058e+00  3.3824e+00  3.2850e+00
 ...
~~~

Using this dataset, we can skip the whole creating our own vectors:

~~~{.python caption=""}

import spacy
import numpy as np

# Load a large English model with word vectors included
nlp = spacy.load('en_core_web_lg')

# Access the vector for a specific word
dog_vector = nlp('dog').vector
bulldog_vector = nlp('bulldog').vector
shoe_vector = nlp('shoe').vector
brick_vector = nlp('brick').vector

def cosine_similarity(vector_a, vector_b):
    dot_product = np.dot(vector_a, vector_b)
    magnitude_a = np.linalg.norm(vector_a)
    magnitude_b = np.linalg.norm(vector_b)
    return dot_product / (magnitude_a * magnitude_b)

similarity_dog_bulldog = cosine_similarity(dog_vector, bulldog_vector)
similarity_shoe_brick = cosine_similarity(shoe_vector, brick_vector)
~~~

~~~
Dog, Bulldog similarity: 0.6215080618858337
shoe, brick similarity: 0.301258385181427
~~~

> We see that `dog` is over twice as related to `bulldog` as `shoe` is to `brick`. This seems vaguely right to me. Surprisingly, though, 'dog' is closer to 'cat' than to 'bulldog,' but this will work for our purposes.

Reviewer: ```suggestion
We see that `dog` is over twice as related to `bulldog` as `shoe` is to `brick`. This seems vaguely right to me. Surprisingly, though, `dog` is closer to `cat` than to `bulldog`, but this will work for our purposes.
```

## Text Embeddings

So now we can do related words but in the real world it would be great to extend this to whole sentences, or titles or even full documents. The simple way to do this might be find the vector of each word in the document and then combine these vectors.

> We can use this, but there are some issues. The primary problem is that writing is complex. The meaning of a sentence is not a combination of the meaning of the various words. "I like dogs" and "I hate dogs" mean the opposite, but combining the weights of individual vectors will end up very close to each other since all but 1 word is precisely the same. Meanwhile, a sentence like "You love dogs" will end up further away because of the difference between "You and "I"

Reviewer: ```suggestion
We can use this, but there are some issues. The primary problem is that writing is complex. The meaning of a sentence is not a combination of the meaning of the various words. "I like dogs" and "I hate dogs" mean the opposite, but combining the weights of individual vectors will end up very close to each other since all but 1 word is precisely the same. Meanwhile, a sentence like "You love dogs" will end up further away because of the difference between "You" and "I"
```
